In [1]:
from numpy.random import randn
import numpy as np
import pandas as pd
np.random.seed(123)
import os
import matplotlib.pyplot as plt
plt.rc('figure', figsize=(10, 6))
np.set_printoptions(precision=4)

# **Step 1: Data Collection**

There are 2 data sets:
* dengue_features_train.csv
* dengue_labels_train.csv

In [2]:
# Reading data:
dengue = pd.read_table('dengue_features_train.csv', sep=',', header=0, delimiter=None, index_col=None, doublequote=True, engine='python')
dengue.head(20)

city  year  weekofyear week_start_date   ndvi_ne   ndvi_nw   ndvi_se  \
0    sj  1990          18      1990-04-30  0.122600  0.103725  0.198483   
1    sj  1990          19      1990-05-07  0.169900  0.142175  0.162357   
2    sj  1990          20      1990-05-14  0.032250  0.172967  0.157200   
3    sj  1990          21      1990-05-21  0.128633  0.245067  0.227557   
4    sj  1990          22      1990-05-28  0.196200  0.262200  0.251200   
5    sj  1990          23      1990-06-04       NaN  0.174850  0.254314   
6    sj  1990          24      1990-06-11  0.112900  0.092800  0.205071   
7    sj  1990          25      1990-06-18  0.072500  0.072500  0.151471   
8    sj  1990          26      1990-06-25  0.102450  0.146175  0.125571   
9    sj  1990          27      1990-07-02       NaN  0.121550  0.160683   
10   sj  1990          28      1990-07-09  0.192875  0.082350  0.191943   
11   sj  1990          29      1990-07-16  0.291600  0.211800  0.301200   
12   sj  1990          30      1990-07-23  0.150567  0.171700  0.226900   
13   sj  1990          31      1990-07-30       NaN  0.247150  0.379700   
14   sj  1990          32      1990-08-06       NaN  0.064333  0.164443   
15   sj  1990          33      1990-08-13       NaN  0.128033  0.206957   
16   sj  1990          34      1990-08-20  0.190233  0.168800  0.167657   
17   sj  1990          35      1990-08-27  0.252900  0.330750  0.264171   
18   sj  1990          36      1990-09-03  0.235400  0.200025  0.283817   
19   sj  1990          37      1990-09-10  0.127967  0.437100  0.123400   

     ndvi_sw  precipitation_amt_mm  reanalysis_air_temp_k  ...  \
0   0.177617                 12.42             297.572857  ...   
1   0.155486                 22.82             298.211429  ...   
2   0.170843                 34.54             298.781429  ...   
3   0.235886                 15.36             298.987143  ...   
4   0.247340                  7.52             299.518571  ...   
5   0.181743                  9.58             299.630000  ...   
6   0.210271                  3.48             299.207143  ...   
7   0.133029                151.12             299.591429  ...   
8   0.123600                 19.32             299.578571  ...   
9   0.202567                 14.41             300.154286  ...   
10  0.152929                 22.27             299.512857  ...   
11  0.280667                 59.17             299.667143  ...   
12  0.214557                 16.48             299.558571  ...   
13  0.381357                 32.66             299.862857  ...   
14  0.138857                 28.80             300.391429  ...   
15  0.168243                 90.75             299.958571  ...   
16  0.172286                 32.40             300.332857  ...   
17  0.284314                 40.94             300.118571  ...   
18  0.230443                 28.86             300.530000  ...   
19  0.148283                 64.56             300.674286  ...   

    reanalysis_precip_amt_kg_per_m2  reanalysis_relative_humidity_percent  \
0                             32.00                             73.365714   
1                             17.94                             77.368571   
2                             26.10                             82.052857   
3                             13.90                             80.337143   
4                             12.20                             80.460000   
5                             26.49                             79.891429   
6                             38.60                             82.000000   
7                             30.00                             83.375714   
8                             37.51                             82.768571   
9                             28.40                             81.281429   
10                            43.72                             81.467143   
11                            40.90                             82.144286   
12                            42.53    

In [3]:
dengue_labels = pd.read_table('dengue_labels_train.csv', sep=',', header=0, delimiter=None, index_col=None, doublequote=True, engine='python')
dengue_labels

city  year  weekofyear  total_cases
0      sj  1990          18            4
1      sj  1990          19            5
2      sj  1990          20            4
3      sj  1990          21            3
4      sj  1990          22            6
...   ...   ...         ...          ...
1451   iq  2010          21            5
1452   iq  2010          22            8
1453   iq  2010          23            1
1454   iq  2010          24            1
1455   iq  2010          25            4

[1456 rows x 4 columns]

# **Step 2: Cleaning Data**

Cleaning Data involes the following key steps:
* 2.a. Understand the data
* 2.b. Handling missing data
* 2.c. Transforming data into meaningful indicators and measures

## 2.a. Understanding the Data

After importing the data, examine the meta data of the data set. 

In [4]:
dengue.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1456 entries, 0 to 1455
Data columns (total 24 columns):
city                                     1456 non-null object
year                                     1456 non-null int64
weekofyear                               1456 non-null int64
week_start_date                          1456 non-null object
ndvi_ne                                  1262 non-null float64
ndvi_nw                                  1404 non-null float64
ndvi_se                                  1434 non-null float64
ndvi_sw                                  1434 non-null float64
precipitation_amt_mm                     1443 non-null float64
reanalysis_air_temp_k                    1446 non-null float64
reanalysis_avg_temp_k                    1446 non-null float64
reanalysis_dew_point_temp_k              1446 non-null float64
reanalysis_max_air_temp_k                1446 non-null float64
reanalysis_min_air_temp_k                1446 non-null float64
reanalysis_precip

**Observations:**
* Total of 1456 records in the data set
* 20 out of the 24 data columns have NaN values. **See Step 2.b. for approaches taken to handle missing data.**
* 'year', 'weekofyear', 'week_start_date' data type may need to be converted to the appropriate time series data types **[Action]**

In [5]:
dengue_labels.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1456 entries, 0 to 1455
Data columns (total 4 columns):
city           1456 non-null object
year           1456 non-null int64
weekofyear     1456 non-null int64
total_cases    1456 non-null int64
dtypes: int64(3), object(1)
memory usage: 45.6+ KB


**Observations:**
* Total of 1456 records in the data set
* No columns have NaN values.
* 'year', 'weekofyear', 'week_start_date' data type may need to be converted to the appropriate time series data types **[Action]**
* 'total_cases' need to be joined to the degue data set. **See Step 2.c. for data transformation steps.**

## 2.b. Handling Missing Data

In the dengue dataset, 20 out of the 24 data columns contain missing data as the number of records in the data column is less than the total number of records in the entire data set. These are the approaches taken to relabel missing data:

* ndvi_ne                                  > 1262 non-null float64 > Pixel northeast of city centroid >> TBD/Mean
* ndvi_nw                                  > 1404 non-null float64 > Pixel northwest of city centroid >> TBD/Mean
* ndvi_se                                  > 1434 non-null float64 > Pixel southeast of city centroid >> TBD/Mean
* ndvi_sw                                  > 1434 non-null float64 > Pixel southeast of city centroid >> TBD/Mean
* precipitation_amt_mm                     > 1443 non-null float64 >> Mean
* reanalysis_air_temp_k                    > 1446 non-null float64 >> Mean
* reanalysis_avg_temp_k                    > 1446 non-null float64 >> Mean
* reanalysis_dew_point_temp_k              > 1446 non-null float64 >> Mean
* reanalysis_max_air_temp_k                > 1446 non-null float64 >> Mean
* reanalysis_min_air_temp_k                > 1446 non-null float64 >> Mean
* reanalysis_precip_amt_kg_per_m2          > 1446 non-null float64 >> Mean
* reanalysis_relative_humidity_percent     > 1446 non-null float64 >> Mean
* reanalysis_sat_precip_amt_mm             > 1443 non-null float64 >> Mean
* reanalysis_specific_humidity_g_per_kg    > 1446 non-null float64 >> Mean
* reanalysis_tdtr_k                        > 1446 non-null float64 >> Mean
* station_avg_temp_c                       > 1413 non-null float64 >> Mean
* station_diur_temp_rng_c                  > 1413 non-null float64 >> Mean
* station_max_temp_c                       > 1436 non-null float64 >> Mean
* station_min_temp_c                       > 1442 non-null float64 >> Mean
* station_precip_mm                        > 1434 non-null float64 >> Mean

In [6]:
#Extract the column names from the df
dengue.columns

Index(['city', 'year', 'weekofyear', 'week_start_date', 'ndvi_ne', 'ndvi_nw',
       'ndvi_se', 'ndvi_sw', 'precipitation_amt_mm', 'reanalysis_air_temp_k',
       'reanalysis_avg_temp_k', 'reanalysis_dew_point_temp_k',
       'reanalysis_max_air_temp_k', 'reanalysis_min_air_temp_k',
       'reanalysis_precip_amt_kg_per_m2',
       'reanalysis_relative_humidity_percent', 'reanalysis_sat_precip_amt_mm',
       'reanalysis_specific_humidity_g_per_kg', 'reanalysis_tdtr_k',
       'station_avg_temp_c', 'station_diur_temp_rng_c', 'station_max_temp_c',
       'station_min_temp_c', 'station_precip_mm'],
      dtype='object')

In [7]:
cols=['ndvi_ne','ndvi_nw', 'ndvi_se','ndvi_sw','precipitation_amt_mm', 'reanalysis_air_temp_k',
       'reanalysis_avg_temp_k', 'reanalysis_dew_point_temp_k',
       'reanalysis_max_air_temp_k', 'reanalysis_min_air_temp_k',
       'reanalysis_precip_amt_kg_per_m2',
       'reanalysis_relative_humidity_percent', 'reanalysis_sat_precip_amt_mm',
       'reanalysis_specific_humidity_g_per_kg', 'reanalysis_tdtr_k',
       'station_avg_temp_c', 'station_diur_temp_rng_c', 'station_max_temp_c',
       'station_min_temp_c', 'station_precip_mm']
dengue[cols]=dengue[cols].fillna(dengue.mean().iloc[0])
dengue.isnull().sum()

city                                     0
year                                     0
weekofyear                               0
week_start_date                          0
ndvi_ne                                  0
ndvi_nw                                  0
ndvi_se                                  0
ndvi_sw                                  0
precipitation_amt_mm                     0
reanalysis_air_temp_k                    0
reanalysis_avg_temp_k                    0
reanalysis_dew_point_temp_k              0
reanalysis_max_air_temp_k                0
reanalysis_min_air_temp_k                0
reanalysis_precip_amt_kg_per_m2          0
reanalysis_relative_humidity_percent     0
reanalysis_sat_precip_amt_mm             0
reanalysis_specific_humidity_g_per_kg    0
reanalysis_tdtr_k                        0
station_avg_temp_c                       0
station_diur_temp_rng_c                  0
station_max_temp_c                       0
station_min_temp_c                       0
station_pre

## 2.c. Transforming Data

1. Joining Case count to Dengue data set
2. Converting time series data types

In [8]:
dengue['Week_Key']=dengue['city']+'_'+dengue['year'].astype(str)+'_'+dengue['weekofyear'].astype(str)
dengue_labels['Week_Key']=dengue_labels['city']+'_'+dengue_labels['year'].astype(str)+'_'+dengue_labels['weekofyear'].astype(str)

In [9]:
dengue_cases = pd.merge(dengue, dengue_labels,how='inner',on='Week_Key')
dengue_cases.head(20)

city_x  year_x  weekofyear_x week_start_date      ndvi_ne   ndvi_nw  \
0      sj    1990            18      1990-04-30     0.122600  0.103725   
1      sj    1990            19      1990-05-07     0.169900  0.142175   
2      sj    1990            20      1990-05-14     0.032250  0.172967   
3      sj    1990            21      1990-05-21     0.128633  0.245067   
4      sj    1990            22      1990-05-28     0.196200  0.262200   
5      sj    1990            23      1990-06-04  2001.031593  0.174850   
6      sj    1990            24      1990-06-11     0.112900  0.092800   
7      sj    1990            25      1990-06-18     0.072500  0.072500   
8      sj    1990            26      1990-06-25     0.102450  0.146175   
9      sj    1990            27      1990-07-02  2001.031593  0.121550   
10     sj    1990            28      1990-07-09     0.192875  0.082350   
11     sj    1990            29      1990-07-16     0.291600  0.211800   
12     sj    1990            30      1990-07-23     0.150567  0.171700   
13     sj    1990            31      1990-07-30  2001.031593  0.247150   
14     sj    1990            32      1990-08-06  2001.031593  0.064333   
15     sj    1990            33      1990-08-13  2001.031593  0.128033   
16     sj    1990            34      1990-08-20     0.190233  0.168800   
17     sj    1990            35      1990-08-27     0.252900  0.330750   
18     sj    1990            36      1990-09-03     0.235400  0.200025   
19     sj    1990            37      1990-09-10     0.127967  0.437100   

     ndvi_se   ndvi_sw  precipitation_amt_mm  reanalysis_air_temp_k  ...  \
0   0.198483  0.177617                 12.42             297.572857  ...   
1   0.162357  0.155486                 22.82             298.211429  ...   
2   0.157200  0.170843                 34.54             298.781429  ...   
3   0.227557  0.235886                 15.36             298.987143  ...   
4   0.251200  0.247340                  7.52             299.518571  ...   
5   0.254314  0.181743                  9.58             299.630000  ...   
6   0.205071  0.210271                  3.48             299.207143  ...   
7   0.151471  0.133029                151.12             299.591429  ...   
8   0.125571  0.123600                 19.32             299.578571  ...   
9   0.160683  0.202567                 14.41             300.154286  ...   
10  0.191943  0.152929                 22.27             299.512857  ...   
11  0.301200  0.280667                 59.17             299.667143  ...   
12  0.226900  0.214557                 16.48             299.558571  ...   
13  0.379700  0.381357                 32.66             299.862857  ...   
14  0.164443  0.138857                 28.80             300.391429  ...   
15  0.206957  0.168243                 90.75             299.958571  ...   
16  0.167657  0.172286                 32.40             300.332857  ...   
17  0.264171  0.284314                 40.94             300.118571  ...   
18  0.283817  0.230443                 28.86             300.530000  ...   
19  0.123400  0.148283                 64.56             300.674286  ...   

    station_avg_temp_c  station_diur_temp_rng_c  station_max_temp_c  \
0            25.442857                 6.900000                29.4   
1            26.714286                 6.371429                31.7   
2            26.714286                 6.485714                32.2   
3            27.471429                 6.771429                33.3   
4            28.942857                 9.371429                35.0   
5            28.114286                 6.942857                34.4   
6            27.414286                 6.771429                32.2   
7            28.371429                 7.685714                33.9   
8            28.328571                 7.385714                33.9   
9            28.328571                 6.514286                33.9   
10           27.557143                 7.157143                31.7   

In [10]:
dengue_cases['total_cases'].sum()

35927

In [11]:
dengue_cases_pivot=dengue_cases.pivot_table('total_cases',index='year_x',columns='city_x',aggfunc='sum',margins=True)
dengue_cases_pivot

city_x      iq       sj    All
year_x                        
1990       NaN    866.0    866
1991       NaN   2801.0   2801
1992       NaN   2371.0   2371
1993       NaN   1156.0   1156
1994       NaN   6533.0   6533
1995       NaN   1333.0   1333
1996       NaN    855.0    855
1997       NaN   1454.0   1454
1998       NaN   4595.0   4595
1999       NaN   1633.0   1633
2000       4.0    492.0    496
2001      14.0   1463.0   1477
2002     704.0    452.0   1156
2003     114.0    949.0   1063
2004     638.0    610.0   1248
2005     311.0   1805.0   2116
2006     436.0    550.0    986
2007     365.0   1969.0   2334
2008     801.0    106.0    907
2009     324.0      NaN    324
2010     223.0      NaN    223
All     3934.0  31993.0  35927